Global Styles: Refactor wp_add_inline_style() to use HTML API.

dmsnell opened the pull request:
The existing `wp_add_inline_style()` code attempts to ensure that the provided code doesn't include STYLE tags, but this is difficult to do. In this change, the HTML API is employed to ensure that the HTML tags are properly detected.

With the ability to detect these more carefully, new options are available to the function when things _go wrong_. This makes the attempt to auto-detect when someone has provided a `<style>`-wrapped input by mistake, and will unwrap that properly. Any other defects are passed transparently into the styles.

Ticket: https://core.trac.wordpress.org/ticket/61828

dmsnell committed:
Global Styles: Refactor wp_add_inline_style() to use HTML API.

dmsnell committed:
Fixup to HTML API Code.

ramonjd committed:
Add since annotation

ramonjd committed:
Adding nested tag tests for wp_add_inline_style()

ramonjd reviewed (commented):
Thanks a lot for working on this @dmsnell 

I've tested this with theme.json, and also adding custom CSS in the site editor (global styles). 

Everything works as expected.

✅  "valid" custom CSS works as expected, including nested rules and other selector manifestations
✅ Other inline styles, such as `core-block-supports-inline-css` have no regressions

I added some tests to ensure current functionality is preserved (stripping balanced style tags and injected content), and also testing the new behaviour - I hope you don't mind!

Also, when adding inline CSS that `wp_add_inline_style()` doesn't like, we'll know 😄 

<img width="775" alt="Screenshot 2024-08-06 at 3 05 03 PM" src="https://github.com/user-attachments/assets/2438b381-d5fd-4b34-b2d2-ee61ea3ae9d6">


Do you think the changes to the HTML API needs testing as well?

peterwilsoncc commented on tests/phpunit/tests/dependencies/styles.php:
```diff
@@ -628,4 +628,54 @@ public function test_wp_maybe_inline_styles_no_path() {
+		$this->assertSame( $expected, get_echo( 'wp_print_styles' ), "Inline styles - $styles - do not match expected values." );
+	}
+
+	public function data_styles_with_nested_tags() {
```
```suggestion
	/*
	 * Data Provider
	 *
	 * @returns array[]
	 */
	public function data_wp_add_inline_style_with_nested_style_tags() {
```

peterwilsoncc commented on tests/phpunit/tests/dependencies/styles.php:
```diff
@@ -628,4 +628,54 @@ public function test_wp_maybe_inline_styles_no_path() {
+				"<link rel='stylesheet' id='handle-css' href='http://example.com?ver=1' type='text/css' media='all' />\n<style id='handle-inline-css' type='text/css'>\n\\3c\\2fstyle><script>alert(0)</script><style>\n</style>\n",
+				'</style><script>alert(0)</script><style>',
+			),
+			array(
```
```suggestion
			'closed_and_reopened' => array(
```

And names for each of the others too. Makes it easier to figure out what happens during failures.

peterwilsoncc commented on tests/phpunit/tests/dependencies/styles.php:
```diff
@@ -628,4 +628,54 @@ public function test_wp_maybe_inline_styles_no_path() {
+	 * Tests detecting nested tags in inline styles.
+	 *
+	 * @ticket 61828
+	 * @dataProvider data_styles_with_nested_tags
```
```suggestion
	 * @dataProvider data_wp_add_inline_style_with_nested_style_tags
```

Convention is for data provideres to use the same name with test replaced by data

peterwilsoncc reviewed (commented):
Some quick notes but will come back to actually test the code.

ramonjd committed:
Apply review feedback: rename and add keys for data provider

ramonjd commented:
Thanks @peterwilsoncc - updated!

swissspidy commented:
FWIW, the same goes for `wp_add_inline_script`, where this was first introduced.

IMHO this is overly defensive programming and overkill for this use case.

In these functions we do a simple regex check _out of courtesy_ to tell developers they are doing it wrong. If they added some weird markup that we don't detect, they deserve to get yelled at by the browser or whatever

dmsnell commented:
@swissspidy I added the fallback conditions just because they were easy and because it wasn't clear what the right end goal should be. if you think we ought instead to truncate the provided content once it exists the STYLE context that would be fine too.

For example, if provided `* {}</style><p>Done!</p>` then we could simply send along `* {}` and not do any more work. The only real condition I considered was that many people might send `<style>* {}</style>` and it seemed reasonable to auto-correct this.

ramonjd commented:
> The only real condition I considered was that many people might send <script>runAThing();</script> and it seemed reasonable to auto-correct this.

I agree, and this was the genesis of this PR. Why not tighten the checks up a little, even if out of courtesy? 


> If they added some weird markup that we don't detect, they deserve to get yelled at by the browser or whatever

Even when that weird markup comes from external sources, such as a theme's custom CSS?

dmsnell committed:
Merge remote-tracking branch 'upstream/trunk' into refactor/update-add-inline-style

dmsnell committed:
Adjust comments, formatting, test names.

swissspidy commented:
The current code in core was not meant to magically fix a dev‘s broken inline css/js. We could actually even consider removing that logic with the doing_it_wrong .

dmsnell committed:
Don't remove that thing.

dmsnell commented:
I've updated this by adjusting the comments and test names. Happy to have someone revert if they don't like the slightly-more descriptive test names. I've also moved the interpolated `$script` var in the test failure output to the end after a `:` to more clearly separate it from the error message.

dmsnell commented:
> Why not tighten the checks up a little, even if out of courtesy?

you definitely won't hear me shouting this virtue 😉

IMO each case warrants its own review. here, we can reliably detect this condition and resolve it, so it seemed fine. whether we want to is more about what we allow and which habits we want to discourage.

leaving it in and taking it out are equally easy to do

ramonjd commented:
> IMO each case warrants its own review. here, we can reliably detect this condition and resolve it, so it seemed fine. whether we want to is more about what we allow and which habits we want to discourage.

Fair point. 

My thinking comes from the following assumptions:

-  the current code attempts a code replace, so I'm assuming this should be retained in some fashion for backwards compatibility. (This PR does that).
- It's possible to add scripts that do things in a theme.json's `"css"` field. A theme can be installed by any user, whether they've had a hand in producing the code or not. We should therefore tighten up the check to prevent it.

Everything else I'd defer to you folks for advice.

dmsnell commented:
The HTML API update has been extracted into https://github.com/WordPress/wordpress-develop/pull/7150

dmsnell committed:
Merge branch 'trunk' into refactor/update-add-inline-style

dmsnell commented:
The HTML API update has been merged into `trunk` and `trunk` has been merged into this branch.
